Lib/test/test_patma.py

```diff
@@ -2886,6 +2893,86 @@ class B(A): ...
                         h = 1
                 self.assertEqual(h, 1)
 
+    def test_patma_union_type(self):
+        IntOrStr = int | str
+        w = None
+        match 0:
+            case IntOrStr():
+                w = 0
+        self.assertEqual(w, 0)
+
+    def test_patma_union_no_match(self):
+        StrOrBytes = str | bytes
+        w = None
+        match 0:
+            case StrOrBytes():
+                w = 0
+        self.assertIsNone(w)
+
+    def test_union_type_positional_subpattern(self):
+        IntOrStr = int | str
+        w = None
+        match 0:
+            case IntOrStr(y):
+                w = y
+        self.assertEqual(w, 0)
+
+    def test_union_type_keyword_subpattern(self):
+        EitherPoint = Point | Point3D
+        p = Point(x=1, y=2)
+        w = None
+        match p:
+            case EitherPoint(x=1, y=2):
+                w = 0
+        self.assertEqual(w, 0)
+
+
+    def test_patma_union_arg(self):
+        p = Point(x=1, y=2)
+        IntOrStr = int | str
+        w = None
+        match p:
+            case Point(IntOrStr(), IntOrStr()):
+                w = 0
+        self.assertEqual(w, 0)
+
+    def test_patma_union_kwarg(self):
+        p = Point(x=1, y=2)
+        IntOrStr = int | str
+        w = None
+        match p:
+            case Point(x=IntOrStr(), y=IntOrStr()):
+                w = 0
+        self.assertEqual(w, 0)
+
+    def test_patma_union_arg_no_match(self):
+        p = Point(x=1, y=2)
+        StrOrBytes = str | bytes
+        w = None
+        match p:
+            case Point(StrOrBytes(), StrOrBytes()):
+                w = 0
+        self.assertIsNone(w)
+
+    def test_patma_union_kwarg_no_match(self):
+        p = Point(x=1, y=2)
+        StrOrBytes = str | bytes
+        w = None
+        match p:
+            case Point(x=StrOrBytes(), y=StrOrBytes()):
+                w = 0
+        self.assertIsNone(w)
+
+    def test_union_type_match_second_member(self):
+        EitherPoint = Point | Point3D
```

Comment:
Please, test `PointGeneric[T] | Whatever...` alias

Reply:
What exactly do we want to test here?

A more interesting case might be what happens in `int | list[int]` vs `list[int] | int`. If we try to match `1`, then with the first alias we'd match, with the second we'd get a type error with the current implementation. 

If that should be avoided, and we want the same result in both cases, the implementation needs to be updated to do a try-except.

Reply:
The current behavior mirrors that of `isinstance`:

```python
>>> t = int | list[int]
>>> isinstance(1, t)
True
>>> t = list[int] | int
>>> isinstance(1, t)
Traceback (most recent call last):
  File "<python-input-12>", line 1, in <module>
    isinstance(1, t)
    ~~~~~~~~~~^^^^^^
TypeError: isinstance() argument 2 cannot be a parameterized generic
```

Maybe a test that tests exactly that mirroring is useful.

Reply:
```python
    def test_union_mirrors_isinstance_failure(self):
        t = list[int] | int

        # get the isinstance result
        try:
            reference = isinstance(1, t)
        except TypeError as exc:
            reference = exc

        result = None
        try:
            match 1:
                case t():
                    result = True
        except TypeError as exc:
            result = exc

        self.assertEqual(result, reference)
```

gives:

```
Expected :TypeError('called match pattern must be a class or types.UnionType (got types.GenericAlias)')
Actual   :TypeError('isinstance() argument 2 cannot be a parameterized generic')
```

Maybe ok if I just check `self.assertEqual(type(result), type(reference))` in this case?

Reply:
This also runs fine:

```python
>>> t = list[int]
>>> match 1:
...   case int() | t():
...     print("match")
...
match
```
while this fails:
```python
>>> t = list[int]
>>> match 1:
...   case t():
...     print("match")
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: called match pattern must be a type
```